pkg/trace/writer/trace.go

```diff
@@ -139,6 +144,49 @@ func (w *TraceWriter) Run() {
 	}
 }
 
+func (w *TraceWriter) runSync() {
+	defer close(w.stop)
+	defer close(w.flushChan)
+	for {
+		select {
+		case pkg := <-w.In:
+			w.addSpans(pkg)
+		case <-w.flushChan:
+			// Make sure all pending spans have been buffered
+			w.drainAndFlush()
+			w.wg.Done()
+		case <-w.stop:
+			w.drainAndFlush()
+			return
+		}
+	}
+}
+
+// FlushSync blocks and sends pending payloads when syncMode is true
+func (w *TraceWriter) FlushSync() error {
+	if !w.syncMode {
+		return log.Errorf("FlushSync called on TraceWriter when in async mode")
+	}
+	defer w.report()
+
+	w.wg.Add(1) // Add will be completed in sync loop
+	w.flushChan <- struct{}{}
+	// Wait for all payloads to be submitted to the sender
+	w.wg.Wait()
```

Comment:
I really don't like this approach. Seems like the wrong use for the concept. There is no "group" here :) How about we make `flushChan` of the type `chan chan struct{}` and use it as such:
```go
notify := make(chan struct{}, 1) // we don't need to block the select statement
w.flushChan <- notify
<-notify
```
Should be a simply synchronisation mechanism, and we can get rid of the struct field. In the select statement, you would pick up the channel and make sure to notify back when done.

Reply:
Sounds good to me